server/src/lib/server.rs

```diff
@@ -667,53 +670,88 @@ async fn instance_state_put(
+
+    let (send_ch, mut recv_ch) = mpsc::channel(4);
+
     loop {
+        ws_streams.extend(incoming_ws_streams.drain(..));
+
         let (uart_read, ws_send) = match &cur_output {
             None => (
                 serial.read_source(&mut output, actx).fuse(),
                 Fuse::terminated(),
             ),
             Some(r) => (
                 Fuse::terminated(),
-                ws_sink.send(Message::binary(&output[r.clone()])).fuse(),
+                futures::stream::iter(
+                    ws_sinks
+                        .iter_mut()
+                        .zip(std::iter::repeat(Vec::from(&output[r.clone()]))),
+                )
+                .for_each_concurrent(4, |(ws, bin)| {
+                    ws.send(Message::binary(bin)).map(|_| ())
+                })
+                .fuse(),
             ),
         };
+
         let (ws_recv, uart_write) = match &cur_input {
-            None => (ws_stream.next().fuse(), Fuse::terminated()),
+            None => (
+                futures::stream::iter(ws_streams.iter_mut().enumerate())
+                    .for_each_concurrent(4, |(i, ws)| {
+                        // if we don't `move` below, rustc says that `i` (which is usize: Copy (!))
+                        // is borrowed. but if we move without making this explicit reference here,
+                        // it moves send_ch into the closure.
+                        let ch = &send_ch;
+                        ws.next().then(move |msg| ch.send((i, msg))).map(|_| ())
+                    })
+                    .fuse(),
+                Fuse::terminated(),
+            ),
             Some((data, consumed)) => (
                 Fuse::terminated(),
                 serial.write_sink(&data[*consumed..], actx).fuse(),
             ),
         };
 
+        let recv_ch_fut = recv_ch.recv();
+
         tokio::select! {
             // Poll in the order written
             biased;
 
-            // It's important we always poll the detach channel first
+            // It's important we always poll the close channel first
             // so that a constant stream of incoming/outgoing messages
-            // don't cause us to ignore a detach
-            _ = &mut detach => {
-                info!(log, "Detaching from serial console");
-                let close = CloseFrame {
-                    code: CloseCode::Policy,
-                    reason: Cow::Borrowed("serial console was detached"),
-                };
-                ws_sink.send(Message::Close(Some(close))).await?;
+            // don't cause us to ignore it
+            _ = &mut close_recv => {
+                info!(log, "Terminating serial task");
                 break;
             }
 
+            new_ws = websocks_recv.recv() => {
+                if let Some(ws) = new_ws {
+                    let (ws_sink, ws_stream) = ws.split();
+                    ws_sinks.push(ws_sink);
+                    incoming_ws_streams.push(ws_stream);
```

Comment:
Why not just push directly to `ws_streams` here?

Reply:
:woman_facepalming: punk'd by borrowck once again. once upon a time it was furious that i was modifying `ws_streams` while `ws_recv` was using it